File: java-imperative-ms/domain/usecase/src/main/java/com/maocq/usecase/cases/CasesUseCase.java
package com.maocq.usecase.cases;

import com.maocq.model.account.Account;
import com.maocq.model.account.gateways.AccountRepository;
import com.maocq.model.hello.gateways.HelloRepository;
import lombok.RequiredArgsConstructor;

import java.time.ZonedDateTime;
import java.util.List;
import java.util.Optional;

@RequiredArgsConstructor
public class CasesUseCase {

  private final AccountRepository accountRepository;
  private final HelloRepository helloRepository;

  public Account caseOne(int latency) {
     var x = helloRepository.hello(latency);
     var y = helloRepository.hello(latency);
     var account = accountRepository.findById(4000)
         .orElseThrow(() -> new IllegalStateException("Account not found"));
     var z = helloRepository.hello(latency);
     account.setUpdateDate(ZonedDateTime.now());
     return accountRepository.update(account);
  }

  public List<Account> caseTwo(int latency) {
    var account = accountRepository.findOne();
    var x = helloRepository.hello(latency);
    return account;
  }

  public Optional<Account> caseThree() {
    return accountRepository.findById(4000);
  }
}

File: elixir-ms/lib/domain/usecase/cases/cases_use_case.ex
defmodule ElixirMs.UseCase.CasesUseCase do
  @moduledoc false

  @account_repository Application.compile_env(:elixir_ms, :account_repository)
  @hello_repository Application.compile_env(:elixir_ms, :hello_repository)

  def case_one(latency) do
    with {:ok, _} <- @hello_repository.hello(latency),
         {:ok, _} <- @hello_repository.hello(latency),
         account <- @account_repository.find_by_id(4000),
         {:ok, _} <- @hello_repository.hello(latency) do

      updated_account = %{account | name: inspect(Timex.now)}
      {:ok, @account_repository.update(updated_account)}
    end
  end

  def case_two(latency) do
    with account <- @account_repository.find_one(),
         {:ok, _} <- @hello_repository.hello(latency) do
      {:ok, account}
    end
  end

  def case_three() do
    @account_repository.find_by_id(4000)
  end
end
